// @element-plus/hooks/use-locale/index.js
import { computed, getCurrentInstance, inject, provide, ref, } from 'vue';
import English from '@element-plus/locale/lang/en';
export const useLocaleProps = {
    locale: {
        type: Object,
    },
    i18n: {
        type: Function,
    },
};
export const LocaleInjectionKey = 'ElLocaleInjection';
// this is meant to fix global methods like `ElMessage(opts)`, this way we can inject current locale
// into the component as default injection value.
// refer to: https://github.com/element-plus/element-plus/issues/2610#issuecomment-887965266
let localeObjCache;
export const useLocale = () => {
    const vm = getCurrentInstance();
    const props = vm.props;
    const locale = computed(() => props.locale || English);
    const lang = computed(() => locale.value.name);
    const _translator = (...args) => {
        const [path, option] = args;
        let value;
        const array = path.split('.');
        let current = locale.value;
        for (let i = 0, j = array.length; i < j; i++) {
            const property = array[i];
            value = current[property];
            if (i === j - 1)
                return template(value, option);
            if (!value)
                return '';
            current = value;
        }
    };
    const t = (...args) => {
        var _a;
        return ((_a = props.i18n) === null || _a === void 0 ? void 0 : _a.call(props, ...args)) || _translator(...args);
    };
    const provides = {
        locale,
        lang,
        t,
    };
    // this could be broken if someone tries to do following:
    /**
     * <config-provider :locale="lang1">
     *   <config-provider :locale="lang2">
     *     Something calls modal component.
     *   </config-provider>
     * </config-provider>
     */
    localeObjCache = provides;
    provide(LocaleInjectionKey, provides);
};
function template(str, option) {
    if (!str || !option)
        return str;
    return str.replace(/\{(\w+)\}/g, (_, key) => {
        return option[key];
    });
}
export const useLocaleInject = () => {
    return inject(LocaleInjectionKey, localeObjCache || {
        lang: ref(English.name),
        locale: ref(English),
        t: (...args) => {
            const [path, option] = args;
            let value;
            const array = path.split('.');
            let current = English;
            for (let i = 0, j = array.length; i < j; i++) {
                const property = array[i];
                value = current[property];
                if (i === j - 1)
                    return template(value, option);
                if (!value)
                    return '';
                current = value;
            }
        },
    });
};
